src/pools.jl

```diff
@@ -4,53 +4,84 @@ export Pool, acquire, release, drain!
 
-The `max` argument will limit the number of objects
-that can be acquired at any given time. If the limit has been reached, `acquire` will
-block until an object is returned to the pool via `release`.
+The `limit` argument will limit the number of objects that can be in use at any given time.
+If the limit has been reached, `acquire` will block until an object is released
+via `release`.
 
-By default, `release(pool, obj)` will return the object to the pool for reuse.
-`release(pool)` will return the "permit" to the pool while not returning
-any object for reuse.
+- `release(pool, obj)` will return the object to the pool for reuse.
+- `release(pool)` will decrement the number in use but not return any object for reuse.
+- `drain!` can be used to remove objects that have been returned to the pool for reuse;
+  it does *not* release any objects that are in use.
 
-`drain!` can be used to remove any cached objects for reuse, but it does *not* release
-any active acquires.
+See also `acquire`, `release`, `Pools.limit`, `Pools.in_use`, `Pools.in_pool`, `drain!`.
 """
 mutable struct Pool{K, T}
     lock::Threads.Condition
-    max::Int
+    limit::Int
     cur::Int
     keyedvalues::Dict{K, Vector{T}}
     values::Vector{T}
 
-    function Pool{K, T}(max::Int=4096) where {K, T}
+    function Pool{K, T}(limit::Int=4096) where {K, T}
         T === Nothing && throw(ArgumentError("Pool type can not be `Nothing`"))
-        x = new(Threads.Condition(), max, 0)
+        x = new(Threads.Condition(), limit, 0)
         if K === Nothing
             x.values = T[]
-            safesizehint!(x.values, max)
+            safesizehint!(x.values, limit)
         else
             x.keyedvalues = Dict{K, Vector{T}}()
         end
         return x
     end
 end
 
-Pool{T}(max::Int=4096) where {T} = Pool{Nothing, T}(max)
+Pool{T}(limit::Int=4096) where {T} = Pool{Nothing, T}(limit)
 
 safesizehint!(x, n) = sizehint!(x, min(4096, n))
 
 # determines whether we'll look up object caches in .keyedvalues or .values
 iskeyed(::Pool{K}) where {K} = K !== Nothing
 
+Base.keytype(::Type{<:Pool{K}}) where {K} = K
+Base.keytype(p::Pool) = keytype(typeof(p))
+
+Base.valtype(::Type{<:Pool{<:Any, T}}) where {T} = T
+Base.valtype(p::Pool) = valtype(typeof(p))
+
+"""
+    Pools.limit(pool::Pool) -> Int
+
+Return the maximum number of objects permitted to be in use at the same time.
+See `Pools.in_use(pool)` for the number of objects currently in use.
+"""
+limit(pool::Pool) = Base.@lock pool.lock pool.limit
+
+"""
+    Pools.in_use(pool::Pool) -> Int
+
+Return the number of objects currently in use. Less than or equal to `Pools.limit(pool)`.
+"""
+in_use(pool::Pool) = Base.@lock pool.lock pool.cur
+
+"""
+    Pools.in_pool(pool::Pool) -> Int
+
+Return the number of objects in the pool available for reuse.
+"""
+in_pool(pool::Pool) = Base.@lock pool.lock mapreduce(length, +, values(pool.keyedvalues); init=0)
+in_pool(pool::Pool{Nothing}) = Base.@lock pool.lock length(pool.values)
```

Comment:
Do we need the locking here? IIUC, this trades one race condition (do we read the a field of a struct before someone changes is) for another (do we grab the lock first), not sure if this is worth a potential context switch.

Reply:
i don't think we want to risk `values` being mutated whilst computing `length`, right?

Reply:
IIUC, this simply reads the `count` field on the `Dict` instance, seems safe to me

Reply:
... and for arrays I thought checking their length is racy but safe

Reply:
seems unnecessary to rely on implementation details when we could guarantee safety -- what would be the benefit of not locking (given we expect the computation to take nanoseconds)?

Reply:
oh, for `limit`, yeah i can see that (because there's no public API to change limit, in fact we should probably `const` it)

Reply:
>Reading from a mutable field in a struct from multiple threads should be locked.

Well, I'm asking _which_ bad things could happen if you don't lock. Like, could the array length value only be partially updated? That would be a good reason to lock.

>I don't see how this introduces a new kind of race condition

I meant that grabbing the lock is "racy" in the sense that sometimes it will be the reader who wins the race and sometimes the mutator (so assuming no "partial" values can be read, it won't be a big improvement).

>it doesn't matter if you block on a lock here

I think avoiding work that is not needed is a good principle to follow